source/utilities/file.bas

```diff
@@ -32,6 +32,110 @@ FUNCTION CopyFile& (sourceFile$, destFile$)
+        OPEN "B", #sff%, tmpfile$: PUT #sff%, , compdata$: CLOSE #sff%
+        packed% = 1
+        OPEN "B", #sff%, tmpfile$
+    ELSE
+        packed% = 0
+        OPEN "B", #sff%, file$
+    END IF
+    '--- init variables ---
+    fl& = LOF(sff%)
+    cntL& = INT(fl& / 32)
+    cntV& = INT(cntL& / 8180)
+    cntB& = (fl& - (cntL& * 32))
+    '--- create C-Array ---
+    dff% = FREEFILE
+    OPEN "A", #dff%, tmpdir$ + "embedded.cpp"
+    '--- process LONGs ---
+    tmpI$ = SPACE$(32)
+    FOR vc& = 0 TO cntV&
+        IF vc& = cntV& THEN numL& = (cntL& MOD 8180): ELSE numL& = 8180
+        PRINT #dff%, "static const uint32_t "; handle$; "L"; LTRIM$(STR$(vc&)); "[] = {"
+        PRINT #dff%, "    "; LTRIM$(STR$(numL& * 8)); ","
+        FOR z& = 1 TO numL&
+            GET #sff%, , tmpI$: offI% = 1
+            tmpO$ = "    " + STRING$(88, ","): offO% = 5
+            DO
+                tmpL& = CVL(MID$(tmpI$, offI%, 4)): offI% = offI% + 4
+                MID$(tmpO$, offO%, 10) = "0x" + RIGHT$("00000000" + HEX$(tmpL&), 8)
+                offO% = offO% + 11
+            LOOP UNTIL offO% > 92
+            IF z& < numL& THEN PRINT #dff%, tmpO$: ELSE PRINT #dff%, LEFT$(tmpO$, 91)
+        NEXT z&
+        PRINT #dff%, "};"
+        PRINT #dff%, ""
+    NEXT vc&
+    '--- process remaining BYTEs ---
+    IF cntB& > 0 THEN
+        PRINT #dff%, "static const uint8_t "; handle$; "B[] = {"
+        PRINT #dff%, "    "; LTRIM$(STR$(cntB&)); ","
+        PRINT #dff%, "    ";
+        FOR x% = 1 TO cntB&
+            GET #sff%, , tmpB%%
+            PRINT #dff%, "0x" + RIGHT$("00" + HEX$(tmpB%%), 2);
+            IF x% <> 16 THEN
+                IF x% <> cntB& THEN PRINT #dff%, ",";
+            ELSE
+                IF x% <> cntB& THEN
+                    PRINT #dff%, ","
+                    PRINT #dff%, "    ";
+                END IF
+            END IF
+        NEXT x%
+        PRINT #dff%, ""
+        PRINT #dff%, "};"
+        PRINT #dff%, ""
+    END IF
+    '--- make a read function ---
+    PRINT #dff%, "qbs *GetArrayData_"; handle$; "(qbs *handle)"
+    PRINT #dff%, "{"
+    PRINT #dff%, "    if (!qbs_equal(handle, qbs_new_txt("; AddQuotes$(handle$); "))) {return qbs_new_txt("; MKI$(&H2222); ");}"
+    PRINT #dff%, ""
+    PRINT #dff%, "    qbs  *data = qbs_new("; LTRIM$(STR$(fl&)); ", 1);"
+    PRINT #dff%, "    void *buff = data -> chr;"
+    PRINT #dff%, ""
+    FOR vc& = 0 TO cntV&
+        PRINT #dff%, "    memcpy(buff, &"; handle$; "L"; LTRIM$(STR$(vc&)); "[1], "; handle$; "L"; LTRIM$(STR$(vc&)); "[0] << 2);"
+        IF vc& < cntV& OR cntB& > 0 THEN
+            PRINT #dff%, "    buff += ("; handle$; "L"; LTRIM$(STR$(vc&)); "[0] << 2);"
+        END IF
+    NEXT vc&
+    IF cntB& > 0 THEN
+        PRINT #dff%, "    memcpy(buff, &"; handle$; "B[1], "; handle$; "B[0]);"
+    END IF
+    PRINT #dff%, ""
+    IF packed% THEN
+        PRINT #dff%, "    return func__inflate(data, "; LTRIM$(STR$(LEN(filedata$))); ", 1);"
```

Comment:
`func__inflate()` returns a `tmp` `qbs`. Do we still need to `qbs_free()` that?

```c
qbs *ret = qbs_new(uncompsize, 1);
```

I am asking this because I am working on something and I see a lot of `libqb` stuff creating strings this way and returning it.

The only time, it is not done this way is when the `qbs` is `static` inside the function and allocated only once during the lifetime of the program.

Reply:
Well I'll be honest, I don't know for sure, there's some spooky stuff that goes on with the string allocation and cleanup.

My thinking was that the `data` string that we pass to `func__inflate()` is created earlier in this function using `qbs_new()`, but that's not the string that gets returned - `func__inflate()` calls `qbs_new()` itself and returns a new string. The `data` string is never used again after the `func__inflate()` call and seems like it's a leak 🤷